core/trino-main/src/main/java/io/trino/operator/scalar/ConcatFunction.java

```diff
@@ -31,32 +32,52 @@
 import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
 import static io.trino.spi.type.VarbinaryType.VARBINARY;
 import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.sql.analyzer.TypeSignatureTranslator.parseTypeSignature;
 import static io.trino.util.Reflection.methodHandle;
 import static java.lang.Math.addExact;
 import static java.util.Collections.nCopies;
 
 public final class ConcatFunction
         extends SqlScalarFunction
 {
-    // TODO design new variadic functions binding mechanism that will allow to produce VARCHAR(x) where x < MAX_LENGTH.
-    public static final ConcatFunction VARCHAR_CONCAT = new ConcatFunction(VARCHAR.getTypeSignature(), "Concatenates given strings");
+    public static final ConcatFunction VARCHAR_CONCAT = new ConcatFunction(VARCHAR);
 
-    public static final ConcatFunction VARBINARY_CONCAT = new ConcatFunction(VARBINARY.getTypeSignature(), "concatenates given varbinary values");
+    public static final ConcatFunction VARBINARY_CONCAT = new ConcatFunction(VARBINARY);
 
     private static final int MAX_INPUT_VALUES = 254;
     private static final int MAX_OUTPUT_LENGTH = DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
 
-    private ConcatFunction(TypeSignature type, String description)
+    private ConcatFunction(Type type)
     {
-        super(FunctionMetadata.scalarBuilder()
-                .signature(Signature.builder()
-                        .name("concat")
-                        .returnType(type)
-                        .argumentType(type)
-                        .variableArity()
-                        .build())
-                .description(description)
-                .build());
+        super(buildFunctionMetadata(type));
+    }
+
+    private static FunctionMetadata buildFunctionMetadata(Type type)
+    {
+        if (type == VARCHAR) {
+            return FunctionMetadata.scalarBuilder()
+                    .signature(Signature.builder()
+                            .name("concat")
+                            .longVariable("x", "min(2147483647, TOTAL_LONG_LITERAL_SIZE)")
+                            .returnType(parseTypeSignature("varchar(x)", ImmutableSet.of("x")))
```

Comment:
would it be possible to populate the return type in a functional manner?
like

```
.returnType(argumentTypes -> { .... })
```

I know this would harm equivalence on `Signature` objects (`Signature` defines `equals` and there is no strict equality for functions). However, `Signature` seems to be used as a map key only for non-calculated signatures (at least here https://github.com/trinodb/trino/blob/93e6bc3289aa0ea31fec33c2166671cfda12a20d/core/trino-main/src/main/java/io/trino/operator/ParametricImplementationsGroup.java#L164-L173 ).  So maybe the equivalence-related problem is solvable. 

@dain @martint @losipiuk @sopel39 thoughts?